src/gpl/src/nesterovBase.h

```diff
@@ -61,24 +91,108 @@ class PlacerBase;
 class Instance;
 class Pin;
 class Net;
-
-class GPin;
 class FFT;
+class GCell;
+class GNet;
+
+class GPin
+{
+ public:
+  GPin(Pin* pin, uint);
+  GPin(const std::vector<Pin*>& pins);
+
+  Pin* pin() const;
+  const std::vector<Pin*>& pins() const { return pins_; }
+
+  GCell* gCell() const { return gCell_; }
+  GNet* gNet() const { return gNet_; }
+
+  void setGCell(GCell* gCell);
+  void setGNet(GNet* gNet);
+
+  int cx() const { return cx_; }
+  int cy() const { return cy_; }
+
+  // clear WA(Weighted Average) variables.
+  void clearWaVars();
+
+  void setMaxExpSumX(float maxExpSumX);
+  void setMaxExpSumY(float maxExpSumY);
+  void setMinExpSumX(float minExpSumX);
+  void setMinExpSumY(float minExpSumY);
+
+  float maxExpSumX() const { return maxExpSumX_; }
+  float maxExpSumY() const { return maxExpSumY_; }
+  float minExpSumX() const { return minExpSumX_; }
+  float minExpSumY() const { return minExpSumY_; }
+
+  bool hasMaxExpSumX() const { return hasMaxExpSumX_; }
+  bool hasMaxExpSumY() const { return hasMaxExpSumY_; }
+  bool hasMinExpSumX() const { return hasMinExpSumX_; }
+  bool hasMinExpSumY() const { return hasMinExpSumY_; }
+
+  void setCenterLocation(int cx, int cy);
+  void updateLocation(const GCell* gCell);
+  void updateDensityLocation(const GCell* gCell);
+
+  uint getId() { return id; }
+
+ private:
+  uint id = 0;
+  GCell* gCell_ = nullptr;
+  GNet* gNet_ = nullptr;
```

Comment:
The rest of the code is using shared_ptr but here they are still raw pointers.   It should either be all shared_ptr or one unique_ptr and all others as raw pointers.